Fix zombie processes created by FullTeaching test suite

augustocristian opened the pull request:
This PR is related to #52 and closes #135, solves the remaining problem with the zombie processes, and improves the suite's maintainability and understability.
Contains the following improvements:
- Now all test cases are parametrized and use common methods that set-up and tear-down the browsers and configuration files.
- The second user and student drivers are managed globally, tearing down them when are used in the Chat-videocall cases.
- replyToComment test, previosusly @Disabled and now enabled and passing
- Minor improvements suggested by Lint (e.g reemplace size()==0 by isEmpty()) and some rename of variables in the files that we modify.

javiertuya commented on src/test/java/com/fullteaching/e2e/no_elastest/common/BaseLoggedTest.java:
```diff
@@ -66,7 +66,7 @@ static void setupAll() { // 28 lines
             properties.load(new FileInputStream("src/test/resources/inputs/test.properties"));
         } catch (IOException ex) {
-            ex.printStackTrace(); // Consider logging the exception instead of printing the stack trace
+            log.error("Properties file could not be loaded");
```
As I understand, the test.properties file is a requirement to run the tests. If it can't be loaded, the test should fail

javiertuya commented on src/test/java/com/fullteaching/e2e/no_elastest/common/BaseLoggedTest.java:
```diff
@@ -224,19 +243,21 @@ private void login(BrowserUser user, String userEmail, String userPass,
         } catch (NotLoggedException | ElementNotFoundException e) {
-            // TODO Auto-generated catch block
-            e.printStackTrace();
+            if (e.getClass().isInstance(NotLoggedException.class)){
+            log.error("The user {} was not logged",user.getClientData());}
+            else{
+                log.error("The userName field not found" );
+            }
```
Shouldn't this make the test fail?

On the other hand, is it really necesary to control the exceptions? There are two diffferent named exceptions, each represents a cause of failure, what if execute `userName = getUserName(user, true, APP_URL);` without any exception control, and let JUnit control the exceptions to fail the test with an error?

javiertuya commented on src/test/java/com/fullteaching/e2e/no_elastest/common/BaseLoggedTest.java:
```diff
@@ -347,7 +368,7 @@ protected void waitSeconds(int seconds) {
             Thread.sleep(1000L * seconds);
         } catch (InterruptedException e) {
-            e.printStackTrace();
+            log.error("Thread.sleep interrupted");
```
It is not frequent to have exceptions when doing a sleep, but, why this does not fails the test?

javiertuya commented on src/test/java/com/fullteaching/e2e/no_elastest/common/EdgeUser.java:
```diff
@@ -51,42 +50,31 @@ public EdgeUser(String userName, int timeOfWaitInSeconds, String testName, Strin
+                selenoidOptions.put("testName", testName + "-" + userIdentifier + "-" + dtf.format(now));
+                selenoidOptions.put("enableVideo", true);
+                selenoidOptions.put("enableVNC", true);
+                selenoidOptions.put("name", testName + "-" + userIdentifier);
                 selenoidOptions.put("enableLog", true);
-                selenoidOptions.put("logName ", logName);
-                selenoidOptions.put("videoName", videoName);
-
+                selenoidOptions.put("logName ", String.format("%s%s",baseName.replaceAll("\\s","") , ".log"));
+                selenoidOptions.put("videoName", String.format("%s%s",baseName.replaceAll("\\s","") ,".mp4"));
                 selenoidOptions.put("screenResolution", "1920x1080x24");
-
                 options.setCapability("selenoid:options", selenoidOptions);
-
                 //END CAPABILITIES FOR SELENOID RETORCH
```
This part of selenoid options seems to appear three times, it should be refactored, perhaps in the base class.
Do not do this now, create a new issue for it and annotate the number by replying this comment.

Note: Maybe there are more lines of code that are repeated, e.g. before this commented block.

augustocristian replied:
Done, created issue #146 with a associated branch from this PR

augustocristian replied:
Addressed in  https://github.com/giis-uniovi/retorch-st-fullteaching/tree/146-reduce-the-duplicated-code-in-base-chrome-edge-firefoxuser

javiertuya commented on src/test/java/com/fullteaching/e2e/no_elastest/functional/test/LoggedLinksTests.java:
```diff
@@ -21,39 +19,42 @@
 
-@ExtendWith(SeleniumJupiter.class)
 class LoggedLinksTests extends BaseLoggedTest {
 
 
 
+
+
```
There was three blank lines that should be only one, and you have added two more. This is common in this class.

javiertuya commented on src/test/java/com/fullteaching/e2e/no_elastest/functional/test/UserTest.java:
```diff
@@ -34,22 +35,23 @@ public static Stream<Arguments> data() throws IOException {
             assertTrue(true, "not logged");
         } catch (NotLoggedException | BadUserException e) {
-            e.printStackTrace();
+            log.debug("The user was not logged");
             fail("Not logged");
         } catch (ElementNotFoundException e) {
-            e.printStackTrace();
+            log.debug("The web element used to check the log was not found");
             fail(e.getLocalizedMessage());
         }
```
Check a previous comment, maybe there is no need to handle exceptions here

javiertuya reviewed (commented):
This PR is mixing too many things, and the goal was to fix a problem, not to refactor all tests. Please DO NOT SUBMIT THIS KIND OF PRs again.

Main issues:
- duplicated code
- handling of exceptions in tests

augustocristian commented:
Sorry for including a lot of changes in this PR @javiertuya. When I started to look for the problem with the zombie processes, I found several main causes, so I addressed all:
- Have a set-up method in the parent  method and in the ones that inheritance 
- Not use correctly try with resources
- Not use DriverFactories and release the driver correctly.
- The issue with the agent, that probably was my blame because originally I changed the way that you introduced to me to deploy the agent (using the docker command) to a docker-compose.
So I tried to manage all  and try to include in the same PR the changes to resolve the problem.
The last commits have removed most of the exceptions from the test suite (was a common practice, I don't know if it was due to the non-expertise of the original developer) and removed the two-blank lines. I think that a complete refactoring of the code is needed also, but It implies to move-remove code, and it should introduce noise to the PR.

javiertuya commented:
@augustocristian Some after-merge comments:
- When removing some statements that make an inner block (e.g. if removing a try cach), it is acceptable to temporarily keep the original indentation if the block is large, so that, the changes that are shown in the PR reflect exactly the part of code that was removed. However, if the block is small, I thing that it would be better to correct the indentation of the block.
- I've seen many statements with an inline comment that computes the number of lines or says anything else. These comments must be removed, eg. 
```
            CourseNavigationUtilities.go2Tab(driver, HOME_ICON);//4 lines
            CourseNavigationUtilities.go2Tab(driver, SESSION_ICON);//4lines
            CourseNavigationUtilities.go2Tab(driver, FORUM_ICON);//4lines
            CourseNavigationUtilities.go2Tab(driver, FILES_ICON);//4lines
            CourseNavigationUtilities.go2Tab(driver, ATTENDERS_ICON);//4lines
```
